[NetAppFiles]ANF-3998 Swagger RestAPI for 2020-02-01

audunn opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

audunn committed:
Adding 2020-02-02 API version folder

audunn committed:
Update 2020-02-02 API version to spec

audunn committed:
Update older API's version to spec v. MountTarget type

audunn committed:
update readme

AutorestCI commented:
/azp run automation - sdk

azuresdkci commented:
Can one of the admins verify this patch?

yungezz commented on specification/netapp/resource-manager/Microsoft.NetApp/stable/2019-05-01/netapp.json:
```diff
@@ -1696,7 +1696,7 @@
           "description": "List of mount targets",
           "items": {
-            "$ref": "#/definitions/mountTargetList"
+            "$ref": "#/definitions/mountTarget"
```
this sounds like a Swagger bug fixing, just to confirm. otherwise it will be a breaking change

yungezz replied:
hi @audunn, could you pls confirm this? thanks.

audunn replied:
Yes, this is a correction mountTargets that is returned is a generic list (array) of mountTargets and is more accureatly represented in swagger like this. This aligns with later API versions of the swagger that included this change in a previous pr.

audunn replied:
Note we need to make a correction on this the mountTarget property is actually represented like so:
`"$ref": "#/definitions/mountTargetProperties"`

audunn committed:
update mountTarget property

AutorestCI commented:
/azp run automation - sdk

audunn committed:
Added missing mountTarget properties, ran prettier

AutorestCI commented:
/azp run automation - sdk

audunn committed:
Ran prettier on new examples

AutorestCI commented:
/azp run automation - sdk

audunn committed:
Spelling correction

AutorestCI commented:
/azp run automation - sdk

audunn committed:
Removed BOM from spec file

AutorestCI commented:
/azp run automation - sdk

pilor commented:
>           "201": {

A PUT should return 201 on creation and 200 if the resource was updated

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:1556 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented on specification/netapp/resource-manager/Microsoft.NetApp/stable/2019-11-01/netapp.json:
```diff
@@ -2514,6 +2501,36 @@
+          "type": "string",
+          "description": "The subnet",
+          "example": "1.2.3.4"
+        },
```
are all these new properties readonly?  If so please mark them as such.  If not, they shouldn't be added to an existing api-version

audunn replied:
This is a swagger fix, those properties are and have been part of the response for some time but were missing from the swagger updated to reflect that.

pilor commented:
>       "get": {

a "list" operation should be a POST action

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:1636 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>               "user1",

typically AAD interaction is done via object IDs or other truly unique identifiers.  Are you sure you want to deal in actual usernames here?  Isn't there a concern of someone using this as an attack vector (i.e. taking someone elses username when they leave the company)

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2025 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         },

"id" is a top level property (like "name" and "type") and should not reside inside the property bag

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2885 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>             "$ref": "#/definitions/hourlyScheduleObject"

including the "object" suffix will make the class names awkward in generated sdks.  I would remove it unless you have a specific need for it

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2891 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         }

For state-like properties booleans are very limiting in the future as your API evolves.  Something like "policyState": "Enabled" here (where "Enabled" is one value of a string enum) would give you much more flexibility

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2921 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         "location": {

its rare to have a tracked child resource.  Is it required that this be a tracked resource with its own location + tagS?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2933 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         }

please include the nextLink property and add it to x-ms-pageable.  If you run into problems with the number of entities and need to start paging responses this will prevent you from breaking your clients AND your current backend implementation doesn't need to change

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2977 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>     "snapshotPolicyDetails": {

where is this referenced?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2980 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         },

you can't update any of these properties, they don't need to be in the request

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3040 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>           "$ref": "#/definitions/snapshotPolicyProperties",

the PATCH request should only include the properties that you support patch on.  If that is all of them then this is fine

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3047 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         }

please enhance the description here.  It is unclear what this property does

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3083 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>     "hourlyScheduleObject": {

can a user specify multiple types of scheduled on one snapshot?  If not, what response do they get if they try?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3067 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>           "description": "Indicates which hour snapshot should be taken",

in what timezone?  Should be explicit

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3096 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>           "example": "Sunday,Monday"

Do you support localized day names?  Why not have a string array instead of a csv?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3121 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor commented:
>         "daysOfMonth": {

why string instead of array of ints?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3147 in ad72d3c. [](commit_id = ad72d3cb0edc8f3001c7eccb9504f6827ea3822a, deletion_comment = False)

pilor reviewed (changes requested):
:clock1:

audunn committed:
Address review comments

AutorestCI commented:
/azp run automation - sdk

audunn committed:
Removed snapshotPolicyDetails that was not referenced

AutorestCI commented:
/azp run automation - sdk

audunn commented:
> > ```
> >   "get": {
> > ```
> 
> a "list" operation should be a POST action
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:1636 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor This is an existing service and running version and I believe this needs a backend change internal ticket will be created for this on our end.

audunn commented:
> > ```
> >           "user1",
> > ```
> 
> typically AAD interaction is done via object IDs or other truly unique identifiers. Are you sure you want to deal in actual usernames here? Isn't there a concern of someone using this as an attack vector (i.e. taking someone elses username when they leave the company)
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2025 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor Good idea this is an existing service and running version and I believe this needs a backend change internal ticket will be created for this on our end

audunn commented:
> > ```
> >     }
> > ```
> 
> For state-like properties booleans are very limiting in the future as your API evolves. Something like "policyState": "Enabled" here (where "Enabled" is one value of a string enum) would give you much more flexibility
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2921 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor  I believe this needs a backend change internal ticket will be created for this on our end

audunn committed:
Removed properties from snapshotPolicyPatch that user cannot change and should not be there

AutorestCI commented:
/azp run automation - sdk

audunn commented:
> > ```
> >     }
> > ```
> 
> please include the nextLink property and add it to x-ms-pageable. If you run into problems with the number of entities and need to start paging responses this will prevent you from breaking your clients AND your current backend implementation doesn't need to change
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:2977 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor  I believe this needs a backend change internal ticket will be created for this on our end

audunn commented:
> > ```
> > "hourlyScheduleObject": {
> > ```
> 
> can a user specify multiple types of scheduled on one snapshot? If not, what response do they get if they try?
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3067 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor  yes they can.

audunn commented:
> > ```
> >       "$ref": "#/definitions/snapshotPolicyProperties",
> > ```
> 
> the PATCH request should only include the properties that you support patch on. If that is all of them then this is fine
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:3047 in [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a). [](commit_id = [ad72d3c](https://github.com/Azure/azure-rest-api-specs/commit/ad72d3cb0edc8f3001c7eccb9504f6827ea3822a), deletion_comment = False)

@pilor yes they do

audunn committed:
Removed tracking from snapShotPolicy and fixed description of schedule hour property

AutorestCI commented:
/azp run automation - sdk

audunn commented:
@pilor can you please review, we have addressed some points, see comments,  and noted where changes are required in backend that will be done in future version and internal tickets have been created.

pilor commented on specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:
```diff
@@ -0,0 +1,3205 @@
+        },
+        "usedBytes": {
+          "type": "integer",
+          "description": "Resource size in bytes"
```
I see one of them was updated, please update them all

audunn replied:
updated

audunn committed:
Removed location from snapshotPolicy, updated decription of usedBytes

AutorestCI commented:
/azp run automation - sdk

audunn commented on specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:
```diff
@@ -0,0 +1,3205 @@
+          "type": "integer",
+          "description": "Monthly snapshot count to keep"
+        },
+        "daysOfMonth": {
```
This fits the current live versions behavior, but we will change this to a List in a future version, internal ticket has been created and will be  hopefully be done soon.

audunn commented on specification/netapp/resource-manager/Microsoft.NetApp/stable/2020-02-01/netapp.json:
```diff
@@ -0,0 +1,3205 @@
+      "description": "Snapshot policy Details for create and update",
+      "type": "object",
+      "properties": {
+        "location": {
```
updated

audunn commented:
Draft status due to status of snapshotPolicy in backend

yungezz commented:
> Draft status due to status of snapshotPolicy in backend

Added `Do Not Merge` label. Pls ping me when the service is ready, and the PR is not draft but ready to merge.

audunn commented:
Some parts of this were deemed not ready and removed to be introduced in a later update. We have another PR for this version
https://github.com/Azure/azure-rest-api-specs/pull/9742
We can close this.
